### ISSUE TITLE
Sub-pixel Rounding Impacting Accuracy Between Renders

### ISSUE THREAD

**Issue by david-carruth** [openseadragon/openseadragon#2348]
Hello,

The sub-pixel rounding on each tile is causing accuracy issues when comparing one render of a given tile to a different render. In my use case, I need pixel perfect renders, or nearly pixel perfect. I am using CIEDE2000 to do a color comparison from one render to another at the same location and same zoom level and am seeing a lot of variance. 

Is there a way to force OpenSeadragon to always place tiles on a whole pixel boundary? I understand that this will likely impact panning and zooming for the user, but I am more concerned about tile accuracy between renders. 

I was able to create this functionality by resetting the position of the tile when the canvas is drawn, but I am curious if there is a better way or even a built in way to get this to happen? There is likely a better place to handle this as well, this is just a proof of concept for what I am trying to achieve and it provided pixel perfect renders in most cases.

`drawCanvas: function( context, drawingHandler, scale, translate ) {
        this.position = new OpenSeadragon.Point(Math.floor(this.position.x), Math.floor(this.position.y))
        var position = this.position.times($.pixelDensityRatio),
            size     = this.size.times($.pixelDensityRatio),
            rendered;
            size     = this.size.times($.pixelDensityRatio),
            rendered;`

Any advice or insight into this issue would be greatly appreciated! And apologies in advance if this is already covered in a different issue. I looked around but didn't find anything.

Thanks!

**Comment by iangilman** [comment]
Interesting! I would have thought the subpixel rounding would make things more accurate, but I guess not in this case. 

We already disable subpixel rounding for other reasons when we think the tiles might contain transparency (for instance, if they are PNG). Take a look at the code for `subPixelRoundingForTransparency` to see how that's implemented. I guess maybe we would want the ability to adjust this even for non-transparent tiles. 

If you force the subPixelRoundingForTransparency mechanism to disable subpixel rounding, does that give you the same results as your proof of concept?

**Comment by david-carruth** [comment]
@iangilman Thanks for the quick response. Yes if I force the `subPixelRoundingForTransparency` I get the same results as my proof of concept. I did have to disable the `context.clearRect` call that occurs right after the rounding or else I would get flashing tiles. Also, forcing the rounding causes a jittery experience, like the tiles are drawn first and then rounded so they shift slightly after rendering. Is that expected? If the rounding occurs before render I would think there would be no movement after render but I could be wrong. That could also be due to the fact that forcing this only changes the position and size of the tile at render time, but doesn't actually change those values in the actual tile object as can be seen in the debugger. Not sure how many other places those values are used.

One thing that this does not solve is the accuracy of the tiles at low zoom levels. I would image that is due to the rounding of the actual tile sizes which could change to up or down from render to render, causing whole tiles to be off slightly. Any thoughts on that?

**Comment by iangilman** [comment]
How did you get it to honor the `subPixelRoundingForTransparency`? Did you trick it into thinking your tiles are transparent? That also gets you the clearRect. It seems like maybe we want a version of this that doesn't rely on the transparency flag. 

When you say it's jittery with subPixelRoundingForTransparency, is it the same jitter as when you do the rounding yourself? My experience with the rounding is that it looks less smooth, but that makes sense because it's moving in discrete steps. That's why one of the options is `ONLY_AT_REST`; that allows for smooth movement while things are in motion, but then the rounded positioning once things settle. 

To me, the lack of smoothness doesn't look like the tiles are drawn first and then rounded... That may be a difference in perception or it may be that your version is doing something different than I'm used to. If you want to share a running version, I'm happy to take a look.

**Comment by iangilman** [comment]
The thing you're describing where the rounding occurs only once they get to their final locations is the behavior of `ONLY_AT_REST`. If you just forced the logic, you wouldn't be getting that behavior. I don't think what you're describing is actually what's happening, but I wouldn't be able to know for sure without taking a look at your version. 

You might be able to share a video, but I'm not sure if these kinds of nuances would show up in a video capture.

**Comment by david-carruth** [comment]
Here is my attempt at a video capture. I am not sure if it comes through enough to see or not but as I zoom in and out, the seams between tiles appear, and then once at rest, the tiles "jitter" or shift and the seams become much less apparent. Currently, we are only rounding tile positions if the tile sizes are a full tile. So the jittering is fairly minimal, its just noticeable enough. The only change I have added is 

```
if (tile.size.x % 1 < 0.001 || tile.size.x % 1 > 0.999) {
        tile.position.x = Math.round(tile.position.x);
        tile.position.y = Math.round(tile.position.y);
    }
```

after `tile.position` and `tile.size` are set in `positionTile()`.

Is `ONLY_AT_REST` part of the 3.0.0 release? Or was it introduced later? I appreciate your help with this. 

https://github.com/openseadragon/openseadragon/assets/77745481/a529d71e-6550-4fb1-ac75-cd5044193c4f

**Comment by iangilman** [comment]
Yeah, it's hard for me to tell from this video. Thank you for sharing it, though! 

`ONLY_AT_REST` was added in 3.1.0 as part of `subPixelRoundingForTransparency`. 

Why are you only rounding in those certain cases? Does that improve the jitter, or is it what's causing the jitter? If it's improving the jitter, are you still getting enough rounding to get the accuracy you need? 

Regarding the seams, I wonder if it's this? https://github.com/openseadragon/openseadragon/issues/1722

**Comment by iangilman** [comment]
@david-carruth Fantastic! I'm glad you got it worked out to your satisfaction :)
